File: tree/1305.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func getAllElements(root1 *TreeNode, root2 *TreeNode) []int {
    s1 := make([]*TreeNode, 0)
    s2 := make([]*TreeNode, 0)
    ret := make([]int, 0)
    for root1 != nil {
        s1 = append(s1, root1)
        root1 = root1.Left
    }
    for root2 != nil {
        s2 = append(s2, root2)
        root2 = root2.Left
    }    
    for {
        if len(s1) == 0 || len(s2) == 0 {
            break
        }
        root1 = s1[len(s1)-1]
        root2 = s2[len(s2)-1]
        if root1.Val < root2.Val {
            ret = append(ret, root1.Val)
            s1 = s1[:len(s1)-1]
            root1 = root1.Right
            for root1 != nil {
                s1 = append(s1, root1)
                root1 = root1.Left
            }            
        }else {
            ret = append(ret, root2.Val)
            s2 = s2[:len(s2)-1]
            root2 = root2.Right
            for root2 != nil {
                s2 = append(s2, root2)
                root2 = root2.Left
            }            
        }
    }
    if len(s1) > 0 {
        for {
            if len(s1) == 0 {
                break
            }
            root1 = s1[len(s1)-1]
            ret = append(ret, root1.Val)
            s1 = s1[:len(s1)-1]
            root1 = root1.Right  
            for root1 != nil {
                s1 = append(s1, root1)
                root1 = root1.Left
            }            
        }
    }else {
        for {
            if len(s2) == 0 {
                break
            }
            root2 = s2[len(s2)-1]
            ret = append(ret, root2.Val)
            s2 = s2[:len(s2)-1]
            root2 = root2.Right
            for root2 != nil {
                s2 = append(s2, root2)
                root2 = root2.Left
            }            
        }        
    }

    return ret
}

File: tree/1305.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> getAllElements(TreeNode* root1, TreeNode* root2) {
        std::vector<int> ret;
        if (!root1 && !root2) {
            return ret;
        }
        std::stack<TreeNode*> s1;
        std::stack<TreeNode*> s2;
        while (true) {
            while (root1) {
                s1.push(root1);
                root1 = root1->left;
            }
            while (root2) {
                s2.push(root2);
                root2 = root2->left;
            }
            if (s1.empty() || s2.empty()) {
                break;
            }
            root1 = s1.top();
            root2 = s2.top();
            if (root1->val < root2->val) {
                ret.emplace_back(root1->val);
                s1.pop();
                root1 = root1->right;
                root2 = nullptr;
            } else {
                ret.emplace_back(root2->val);
                s2.pop();
                root1 = nullptr;
                root2 = root2->right;
            }
        }
        if (!s1.empty()) {
            single(s1, ret);   
        } else {
            single(s2, ret);
        }
        return ret;
    }
    
    void single(std::stack<TreeNode*> s, std::vector<int>& ret) {
        TreeNode* root = nullptr;
        while (true) {
            while (root) {
                s.push(root);
                root = root->left;
            }
            if (s.empty()) {
                break;
            }
            root = s.top();
            s.pop();
            ret.emplace_back(root->val);
            root = root->right;
        }
        return;
    }
};
